src/geohash_helper.c

```diff
@@ -100,34 +99,43 @@ uint8_t geohashEstimateStepsByRadius(double range_meters, double lat) {
  * Since this function is currently only used as an optimization, the
  * optimization is not used for very big radiuses, however the function
  * should be fixed. */
-int geohashBoundingBox(double longitude, double latitude, double radius_meters,
-                       double *bounds) {
+int geohashBoundingBox(GeoShape *shape, double *bounds) {
     if (!bounds) return 0;
-
-    bounds[0] = longitude - rad_deg(radius_meters/EARTH_RADIUS_IN_METERS/cos(deg_rad(latitude)));
-    bounds[2] = longitude + rad_deg(radius_meters/EARTH_RADIUS_IN_METERS/cos(deg_rad(latitude)));
-    bounds[1] = latitude - rad_deg(radius_meters/EARTH_RADIUS_IN_METERS);
-    bounds[3] = latitude + rad_deg(radius_meters/EARTH_RADIUS_IN_METERS);
+    double longitude = shape->xy[0];
+    double latitude = shape->xy[1];
+    double height = shape->conversion * (shape->type == CIRCULAR_TYPE ? shape->t.radius : shape->t.r.height/2);
+    double width = shape->conversion * (shape->type == CIRCULAR_TYPE ? shape->t.radius : shape->t.r.width/2);
+
+    const double long_delta = rad_deg(width/EARTH_RADIUS_IN_METERS/cos(deg_rad(latitude)));
+    const double lat_delta = rad_deg(height/EARTH_RADIUS_IN_METERS);
+    bounds[0] = longitude - long_delta;
+    bounds[2] = longitude + long_delta;
+    bounds[1] = latitude - lat_delta;
+    bounds[3] = latitude + lat_delta;
     return 1;
 }
 
 /* Return a set of areas (center + 8) that are able to cover a range query
- * for the specified position and radius. */
-GeoHashRadius geohashGetAreasByRadius(double longitude, double latitude, double radius_meters) {
+ * for the specified position and shape (see geohash.h GeoShape). */
+GeoHashRadius geohashGetAreasByShape(GeoShape *shape) {
     GeoHashRange long_range, lat_range;
     GeoHashRadius radius;
     GeoHashBits hash;
     GeoHashNeighbors neighbors;
     GeoHashArea area;
     double min_lon, max_lon, min_lat, max_lat;
-    double bounds[4];
     int steps;
 
-    geohashBoundingBox(longitude, latitude, radius_meters, bounds);
-    min_lon = bounds[0];
-    min_lat = bounds[1];
-    max_lon = bounds[2];
-    max_lat = bounds[3];
+    geohashBoundingBox(shape, shape->bounds);
+    min_lon = shape->bounds[0];
+    min_lat = shape->bounds[1];
+    max_lon = shape->bounds[2];
+    max_lat = shape->bounds[3];
+
+    double longitude = shape->xy[0];
+    double latitude = shape->xy[1];
+    double radius_meters = shape->conversion * (shape->type == CIRCULAR_TYPE ? shape->t.radius :
+                               shape->t.r.width > shape->t.r.height ? shape->t.r.width/2 : shape->t.r.height/2);
 
     steps = geohashEstimateStepsByRadius(radius_meters,latitude);
```

Comment:
i'm not certain this is the right thing to do.
maybe we need an "geohashEstimateStepsBy**Box**" instead of convert our box to a circle?
can you look into that?

p.s. if we keep this, i think it'll look nicer if the conversion will be done in a separate line after the ternary operator.

Reply:
We use radius to determine the steps, which is an approximate interval of geohash, because we will search for 8 nearby areas, so max(w/2,h/2) is correct.

Reply:
in some cases, maybe if the box is very long and narrow, maybe we can do better (maybe find a totally different number of steps (not even a 3x3 set, maybe a 2x6 set, or 2x8).
but i'm ok to overlook this for now.

Reply:
The geohashcalculateareasbyshapewgs84 function finally excludes areas that do not need to be searched.

```
/* Exclude the search areas that are useless. */
    if (steps >= 2) {
        if (area.latitude.min < min_lat) {
            GZERO(neighbors.south);
            GZERO(neighbors.south_west);
            GZERO(neighbors.south_east);
        }
...
```

Reply:
it excludes areas not covered by the radius we generated, but this radius is not optimal, so in practice it may include areas that are outside our box.
also IIUC this trick with radius can cause us to use non-optimal resolution and be inaccurate.
![image](https://user-images.githubusercontent.com/7045099/101514246-bd0b4c80-3985-11eb-8fb9-ea42bbc7843f.png)
in the above image, we could have been searching only in the 6 small light gray boxes.
but instead, we'll be searching in the 9 larger boxes, and skip the 5 on the top and right raw (so end up searching in the 4 large boxes on the bottom left (the majority of that area is outside the ideal search area (the one painted in light gray)

Reply:
also, i now realize that the way we converted the box to a circle isn't right (a bug):
![image](https://user-images.githubusercontent.com/7045099/101514963-808c2080-3986-11eb-93f7-7262d3a5ac7f.png)
seems that we need to call Pythagoras back to the game!

Reply:
> but instead, we'll be searching in the 9 larger boxes, and skip the 5 on the top and right raw (so end up searching in the 4 large boxes on the bottom left (the majority of that area is outside the ideal search area (the one painted in light gray)

no, for the above picture, we will only search two boxes, the middle one and the one below it.
Notice:  we use box (min_lon, min_lat, max_lon, max_lat) to exclude the useless search area, instead of using the circle calculated by radius, so the 2 boxes in the lower left corner will also be excluded.


> also, i now realize that the way we converted the box to a circle isn't right (a bug):
seems that we need to call Pythagoras back to the game!

Currently max(w/2, h/2) is used, and because we will search 9 nearby areas, the current code is no problem, because the 9 areas must completely include the box, even if some areas are excluded by the box later , It has no effect.

Reply:
@yangbodong22011 thank you. you're right in both case.
I'm still feel a bit worried that some of the logic with the "steps" calculation can go wrong in some edge case due to the fact the radius is smaller than it should be, but i don't have any evidence for that.

Maybe one day we can write a fuzz tester that tests random points around the box corners (inside and outside) of that box, and places a box with random dimensions, and random locations on the glove to verify that it doesn't have bugs. But let's keep that for a separate PR.

Regardless I also still feel that it can be made more efficient (searching the 6 smaller gray boxes in my example rather than two larger ones. but also, let's keep that for some future date.

I think a comment is needed next to where we calculate the inaccurate radius to explain why it's safe, please add that.

Reply:
@oranagra thanks，already updated.